src/RwaUrn.sol

```diff
@@ -26,37 +25,50 @@ contract RwaUrn {
         _;
     }
 
-    VatLike  public vat;
-    JoinLike public gemJoin;
-    JoinLike public daiJoin;
-    address  public fbo;
+    VatAbstract  public vat;
+    GemJoinAbstract public gemJoin;
+    DaiJoinAbstract public daiJoin;
+    address  public fbo; // routing conduit?
 
     // --- init ---
     constructor(address vat_, address gemJoin_, address daiJoin_, address fbo_) public {
-        vat = VatLike(vat_);
-        gemJoin = JoinLike(gemJoin_);
-        daiJoin = JoinLike(daiJoin_);
+        // requires in urn that fbo isn't address(0)
+        vat = VatAbstract(vat_);
+        gemJoin = GemJoinAbstract(gemJoin_);
+        daiJoin = DaiJoinAbstract(daiJoin_);
         fbo = fbo_;
         wards[msg.sender] = 1;
+        DSTokenAbstract(gemJoin.gem()).approve(address(gemJoin), uint(-1));
+        DaiAbstract(daiJoin.dai()).approve(address(daiJoin), uint(-1));
+        VatAbstract(vat_).hope(address(daiJoin));
+    }
+
+    // --- administration ---
+    function file(bytes32 what, address data) external auth {
+        // add require statement ensuring address != 0
+        if (what == "fbo") fbo = data;
+        else revert("RwaUrn/unrecognised-param");
     }
 
     // --- cdp operation ---
     // n.b. DAI can only go to fbo
     function lock(uint256 wad) external operator {
-        gemJoin.gem().transferFrom(msg.sender, address(this), wad);
+        DSTokenAbstract(gemJoin.gem()).transferFrom(address(msg.sender), address(this), wad);
+        // join with address this
         gemJoin.join(address(this), wad);
         vat.frob(gemJoin.ilk(), address(this), address(this), address(this), int(wad), 0);
     }
     function free(uint256 wad) external operator {
         vat.frob(gemJoin.ilk(), address(this), address(this), address(this), -int(wad), 0);
         gemJoin.exit(address(this), wad);
-        gemJoin.gem().transfer(msg.sender, wad);
+        DSTokenAbstract(gemJoin.gem()).transfer(address(msg.sender), wad);
     }
     function draw(uint256 wad) external operator {
         vat.frob(gemJoin.ilk(), address(this), address(this), address(this), 0, int(wad));
-        daiJoin.exit(fbo, wad);
+        daiJoin.exit(address(fbo), wad);
```

Comment:
Not sure why this is being cast to an `address` when it is already of that type.

Reply:
yeah, useless cast, fixed.